### DISCUSSION TITLE
Native.js not packed

### DISCUSSION THREAD

**Discussion by bobbytreed** [webpack/webpack#17382] [Q&A]
# Bug report

Typically when I can't find any information on an issue, it means it's so simple that everyone else knows it, and I'm the slow one.  With that in mind, I apologize in advance for my question.  I'm using lmdb with webpack for a vs code extension.
The code crashes when it tries to grab native.js which was not packed.

let yu = Rl(
          (0, n.dirname)(
            (0, Dl.fileURLToPath)(
              "file:///D:/a/1/s/src/LanguageServer/Client/node_modules/lmdb/native.js"
            )
          ).replace(/dist$/, "")
        );
        function vu(e) {  
You can see, in the built extension code that the path is actually the path built as part of the build pipeline meaning it most likely wasn't packaged.

Webpack lmdb

I just need to know how to handle native.js, or whether I "should". I have found discussions saying that some modules should not be used in some places.

https://www.npmjs.com/package/lmdb

Here are examples of other non-typescript code that I've had to treat differently in webpack.  I'm not sure if this is one of those.

```
			{
				test: /\.ts$/,
				exclude: /node_modules/,
				use: [
					{
						loader: 'ts-loader',
                        options: {
                            compilerOptions: {
                                module: "esnext"
                            }
                        }  
					}
				]                      
			},
            {
                test: /playwright-core.*(md|ttf|apk|exe|png$|ps1$|sh$|html$)/i,
                use: 'raw-loader',
            },
            {
                test: /eslint-(config|plugin)-prettier/i,
                use: 'raw-loader',
            },
            {
                test: /\.node$/,
                loader: "node-loader",
            },
```

**Other relevant information:**
webpack version:  5.0 
Node.js version: 16
Operating System: windows
Additional tools: vs code

**Comment by alexander-akait** [comment]
Please provide reproducible test repo. 

Where you get:
```
let yu = Rl(
(0, n.dirname)(
(0, Dl.fileURLToPath)(
"file:///D:/a/1/s/src/LanguageServer/Client/node_modules/lmdb/native.js"
)
).replace(/dist$/, "")
);
function vu(e) {
```
?

**Comment by bobbytreed** [comment]
I added you to the repository 

https://github.com/bobbytreed/test-extension

The out directory is there where you can find the native.js call out.  

Please excuse the hasty coding.  I was just trying to throw something together.

Thanks!

**Reply by alexander-akait** [reply]
Yeah, you have:
```
let dirName = (0,path__WEBPACK_IMPORTED_MODULE_0__.dirname)((0,url__WEBPACK_IMPORTED_MODULE_1__.fileURLToPath)("file:///home/akait/IdeaProjects/test-extension/node_modules/lmdb/native.js")).replace(/dist$/, '');
```

Original code:
```
let dirName = dirname(fileURLToPath(import.meta.url)).replace(/dist$/, '');
```

But commonjs doesn't support `import.meta.url` (you know), so webpack just replace it on the static value

But good news - `lmdb` has CommonJS:
```
 "exports": {
    ".": {
      "types": {
        "require": "./index.d.cts",
        "import": "./index.d.ts"
      },
      "node": {
        "require": "./dist/index.cjs",
        "import": "./node-index.js"
      },
      "default": "./index.js"
    }
  },
```

But you have `module: "esnext"`, so `typescript` generates `import`s and `webpack` gets `import`, so just use:
```
 resolve: {
  alias: {
    'lmdb': path.resolve(__dirname, 'node_modules/lmdb/dist/index.cjs'),
  },
  modules: ['node_modules'],
  extensions: ['.ts', '.js', '.json'],
  mainFields: ["main", "module"]
},
```

But my recommendation is do not use `module: "esnext"` and `mainFields: ["main", "module"]` (i.e. do not add `module`) when your output is commonjs, because you can faced with such limitations.

When I fixed it in your repo I get:
```
ERROR in ./node_modules/lmdb/dist/index.cjs 2610:41-72
Module not found: Error: Can't resolve 'cbor-x' in '/home/akait/IdeaProjects/test-extension/node_modules/lmdb/dist'
 @ ./extension.ts 1:0-28 4:13-17
```

Installed it I got another one:
```
/home/akait/IdeaProjects/test-extension/out/extension.js:751
  var platformPackage = (packageName[0] == '@' ? '' : '@' + packageName + '/') + packageName + '-' + platform + '-' + arch
                                    ^

TypeError: Cannot read properties of undefined (reading '0')
    at load.path (/home/akait/IdeaProjects/test-extension/out/extension.js:751:37)
    at Object.load [as default] (/home/akait/IdeaProjects/test-extension/out/extension.js:720:25)
    at ./node_modules/lmdb/dist/index.cjs (/home/akait/IdeaProjects/test-extension/out/extension.js:3597:47)
    at __webpack_require__ (/home/akait/IdeaProjects/test-extension/out/extension.js:10301:41)
```

Because of (in `./node_modules/node-gyp-build-optional-packages/index.js`):
```
packageName = require(path.join(dir, 'package.json')).name
```

Maybe you don't want to compile this module or you want to exclude it or something else, I don't know the context, or maybe it's even expected behavior, if you provide more context then I can advise a solution.

**Comment by kriszyp** [comment]
To clarify, in https://github.com/kriszyp/node-gyp-build/blob/ngbop/index.js#L37-L40 is webpack not going into the webpack conditional branch? Or is it that the package.json just can't be bundled correctly with a dynamic path?
